code/game/machinery/ftl_shieldgen.dm

```diff
@@ -67,13 +72,25 @@
 		update_icon()
 		update_physical()
 		return
-	if(power_charge < power_charge_max)		// if there's power available, try to charge
-		var/load = charge_rate		// FUCK SEC
-		power_terminal.power_requested = load
-		power_charge += min((power_charge_max-power_charge), power_terminal.last_power_received * CELLRATE)
-		charging_power = 1
-	else
-		charging_power = 0
+	if(on = 1)
+		if(power_charge < power_charge_max)		// if there's power available, try to charge
+			var/load = charge_rate		// FUCK SEC
+			power_terminal.power_requested = load
+			power_charge += min((power_charge_max-power_charge), power_terminal.last_power_received * CELLRATE)
+			charging_power = 1
+		else
+			charging_power = 0
+
+	if(plasma_charge > 0)
+		plasma_charge -= min(plasma_passiveloss, plasma_charge)
+
+	if(power_charge > 0)
+		power_charge -= min(power_passiveloss, power_charge)
+
+	if(power_charge < 100 || plasma_charge < 5)
+		if(power_charge < 100)
+			plasma_charge = 0
+		drop_physical(1)
```

Comment:
All of this can be under update_physical as that's the proc that handles shielding being up or down